internal/record: panic: queue is full

Seen at https://travis-ci.org/cockroachdb/pebble/jobs/589904276:

```
panic: queue is full
goroutine 11202 [running]:
github.com/cockroachdb/pebble/internal/record.(*syncQueue).push(...)
	/Users/travis/gopath/src/github.com/cockroachdb/pebble/internal/record/log_writer.go:84
github.com/cockroachdb/pebble/internal/record.(*LogWriter).Close(0xc0003f6500, 0x0, 0x0)
	/Users/travis/gopath/src/github.com/cockroachdb/pebble/internal/record/log_writer.go:358 +0x263
github.com/cockroachdb/pebble.(*DB).makeRoomForWrite(0xc00043a700, 0x0, 0x0, 0x0)
	/Users/travis/gopath/src/github.com/cockroachdb/pebble/db.go:996 +0x8dd
github.com/cockroachdb/pebble.(*DB).AsyncFlush(0xc00043a700, 0xc0005fc780, 0x3, 0x20)
	/Users/travis/gopath/src/github.com/cockroachdb/pebble/db.go:853 +0x98
github.com/cockroachdb/pebble.(*DB).Flush(...)
	/Users/travis/gopath/src/github.com/cockroachdb/pebble/db.go:830
github.com/cockroachdb/pebble.TestDBConcurrentCommitCompactFlush.func1(0xc0002b4150, 0xc00043a700, 0xc000114d00, 0x304)
	/Users/travis/gopath/src/github.com/cockroachdb/pebble/db_test.go:848 +0x1f1
created by github.com/cockroachdb/pebble.TestDBConcurrentCommitCompactFlush
	/Users/travis/gopath/src/github.com/cockroachdb/pebble/db_test.go:840 +0x1a9
```

This is a "should never happen" type of error, but looking at the stack trace I'm suspicious that the mechanism to ensure this error should never happen is not accounting for the `syncQueue.push` from `LogWriter.Close`.

--- a/commit.go
+++ b/commit.go
@@ -13,10 +13,6 @@ import (
 	"github.com/cockroachdb/pebble/internal/record"
 )
 
-// The maximum concurrency allowed for commit operations. This limit is
-// enforced by commitPipeline.sem.
-const commitConcurrency = record.SyncConcurrency
-
 // commitQueue is a lock-free fixed-size single-producer, multi-consumer
 // queue. The single producer can enqueue (push) to the head, and consumers can
 // dequeue (pop) from the tail.
@@ -42,7 +38,7 @@ type commitQueue struct {
 	// power of 2. A slot is in use until *both* the tail index has moved beyond
 	// it and the slot value has been set to nil. The slot value is set to nil
 	// atomically by the consumer and read atomically by the producer.
-	slots [commitConcurrency]unsafe.Pointer
+	slots [record.SyncConcurrency]unsafe.Pointer
 }
 
 const dequeueBits = 32
@@ -61,31 +57,29 @@ func (q *commitQueue) pack(head, tail uint32) uint64 {
 }
 
 func (q *commitQueue) enqueue(b *Batch) {
-	for {
-		ptrs := atomic.LoadUint64(&q.headTail)
-		head, tail := q.unpack(ptrs)
-		if (tail+uint32(len(q.slots)))&(1<<dequeueBits-1) == head {
-			// Queue is full.
-			panic("not reached")
-		}
-		slot := &q.slots[head&uint32(len(q.slots)-1)]
-
-		// Check if the head slot has been released by dequeue.
-		for atomic.LoadPointer(slot) != nil {
-			// Another goroutine is still cleaning up the tail, so the queue is
-			// actually still full. We spin because this should resolve itself
-			// momentarily.
-			runtime.Gosched()
-		}
-
-		// The head slot is free, so we own it.
-		atomic.StorePointer(slot, unsafe.Pointer(b))
-
-		// Increment head. This passes ownership of slot to dequeue and acts as a
-		// store barrier for writing the slot.
-		atomic.AddUint64(&q.headTail, 1<<dequeueBits)
-		return
+	ptrs := atomic.LoadUint64(&q.headTail)
+	head, tail := q.unpack(ptrs)
+	if (tail+uint32(len(q.slots)))&(1<<dequeueBits-1) == head {
+		// Queue is full. This should never be reached because commitPipeline.sem
+		// limits the number of concurrent operations.
+		panic("pebble: not reached")
 	}
+	slot := &q.slots[head&uint32(len(q.slots)-1)]
+
+	// Check if the head slot has been released by dequeue.
+	for atomic.LoadPointer(slot) != nil {
+		// Another goroutine is still cleaning up the tail, so the queue is
+		// actually still full. We spin because this should resolve itself
+		// momentarily.
+		runtime.Gosched()
+	}
+
+	// The head slot is free, so we own it.
+	atomic.StorePointer(slot, unsafe.Pointer(b))
+
+	// Increment head. This passes ownership of slot to dequeue and acts as a
+	// store barrier for writing the slot.
+	atomic.AddUint64(&q.headTail, 1<<dequeueBits)
 }
 
 func (q *commitQueue) dequeue() *Batch {
@@ -225,7 +219,10 @@ type commitPipeline struct {
 func newCommitPipeline(env commitEnv) *commitPipeline {
 	p := &commitPipeline{
 		env: env,
-		sem: make(chan struct{}, commitConcurrency),
+		// NB: the commit concurrency is one less than SyncConcurrency because we
+		// have to allow one "slot" for a concurrent WAL rotation which will close
+		// and sync the WAL.
+		sem: make(chan struct{}, record.SyncConcurrency-1),
 	}
 	return p
 }

--- a/commit_test.go
+++ b/commit_test.go
@@ -15,6 +15,7 @@ import (
 
 	"github.com/cockroachdb/pebble/internal/arenaskl"
 	"github.com/cockroachdb/pebble/internal/record"
+	"github.com/cockroachdb/pebble/vfs"
 	"golang.org/x/exp/rand"
 )
 
@@ -151,6 +152,93 @@ func TestCommitPipelineAllocateSeqNum(t *testing.T) {
 	}
 	if s := atomic.LoadUint64(&e.visibleSeqNum); total != s {
 		t.Fatalf("expected %d, but found %d", total, s)
+	}
+}
+
+type syncDelayFile struct {
+	vfs.File
+	waiting chan struct{}
+	done    chan struct{}
+}
+
+func (f *syncDelayFile) Sync() error {
+	select {
+	case f.waiting <- struct{}{}:
+	default:
+	}
+	<-f.done
+	return nil
+}
+
+func TestCommitPipelineWALClose(t *testing.T) {
+	// This test stresses the edge case of N goroutines blocked in the
+	// commitPipeline waiting for the log to sync when we concurrently decide to
+	// rotate and close the log.
+
+	mem := vfs.NewMem()
+	f, err := mem.Create("test-wal")
+	if err != nil {
+		t.Fatal(err)
+	}
+	// syncDelayFile will send to the waiting channel when Sync is called, and
+	// then block on the done channel befor returning from that call.
+	sf := &syncDelayFile{
+		File:    f,
+		waiting: make(chan struct{}, 1),
+		done:    make(chan struct{}),
+	}
+
+	// A basic commitEnv which writes to a WAL.
+	wal := record.NewLogWriter(sf, 0 /* logNum */)
+	testEnv := commitEnv{
+		logSeqNum:     new(uint64),
+		visibleSeqNum: new(uint64),
+		apply: func(b *Batch, mem *memTable) error {
+			return nil
+		},
+		write: func(b *Batch, syncWG *sync.WaitGroup, syncErr *error) (*memTable, error) {
+			_, err := wal.SyncRecord(b.storage.data, syncWG, syncErr)
+			return nil, err
+		},
+	}
+	p := newCommitPipeline(testEnv)
+
+	// Launch N (commitConcurrency) goroutines which each create a batch and
+	// commit it with sync==true. Because of the syncDelayFile, none of these
+	// operations can complete until syncDelayFile.done is closed.
+	errCh := make(chan error, cap(p.sem))
+	for i := 0; i < cap(errCh); i++ {
+		go func(i int) {
+			b := &Batch{}
+			if err := b.LogData([]byte("foo"), nil); err != nil {
+				errCh <- err
+				return
+			}
+			errCh <- p.Commit(b, true /* sync */)
+		}(i)
+	}
+
+	<-sf.waiting
+	// At this point, we know at least one operation is blocked waiting for the
+	// Sync to finish, but most likely all are blocked.
+
+	// Fire off another goroutine which will wait a short period of time and then
+	// released the Sync. There isn't a clean way to do this as the wal.Close()
+	// below will block indefinitely until the Sync is released.
+	go func() {
+		time.Sleep(20 * time.Millisecond)
+		close(sf.done)
+	}()
+
+	// Close the WAL. A "queue is full" panic means that something is broken.
+	if err := wal.Close(); err != nil {
+		t.Fatal(err)
+	}
+	for i := 0; i < cap(errCh); i++ {
+		err := <-errCh
+		if err != nil {
+			t.Fatal(err)
+		}
 	}
 }
 

--- a/internal/record/log_writer.go
+++ b/internal/record/log_writer.go
@@ -34,8 +34,9 @@ const (
 	syncConcurrencyBits = 9
 
 	// SyncConcurrency is the maximum number of concurrent sync operations that
-	// can be performed. Exported as this value also limits the commit
-	// concurrency in commitPipeline.
+	// can be performed. Note that a sync operation is initiated either by a call
+	// to SyncRecord or by a call to Close. Exported as this value also limits
+	// the commit concurrency in commitPipeline.
 	SyncConcurrency = 1 << syncConcurrencyBits
 )
 
@@ -81,7 +82,7 @@ func (q *syncQueue) push(wg *sync.WaitGroup, err *error) {
 	ptrs := atomic.LoadUint64(&q.headTail)
 	head, tail := q.unpack(ptrs)
 	if (tail+uint32(len(q.slots)))&(1<<dequeueBits-1) == head {
-		panic("queue is full")
+		panic("pebble: queue is full")
 	}
 
 	slot := &q.slots[head&uint32(len(q.slots)-1)]
@@ -406,7 +407,11 @@ func (w *LogWriter) SyncRecord(p []byte, wg *sync.WaitGroup, err *error) (int64,
 	}
 
 	offset := w.blockNum*blockSize + int64(w.block.written)
-	return offset, w.err
+	// Note that we don't return w.err here as a concurrent call to Close would
+	// race with our read. That's ok because the only error we could be seeing is
+	// one to syncing for which the caller can receive notification of by passing
+	// in a non-nil err argument.
+	return offset, nil
 }
 
 // Size returns the current size of the file.

